<a href="https://colab.research.google.com/github/v2047/Machine-Leaning-Projects/blob/main/Credit_Card_Fraud_Detection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**About the data:**

Features V1, V2, … V28 are the principal components obtained with PCA, the only features which have not been transformed with PCA are 'Time' and 'Amount'. Feature 'Time' contains the seconds elapsed between each transaction and the first transaction in the dataset. The feature 'Amount' is the transaction Amount, this feature can be used for example-dependant cost-sensitive learning. Feature 'Class' is the response variable and it takes value **1 in case of fraud 0 otherwise**.


Time : time elasped in second

1 - fraudulent transactions

0 - valid transactions

WorkFlow :

**Credit Card Data ------> Data Pre processing -----> Data Analysis ----->Train Test Split -----> ML Model (Logistic) -----> Evaluation**

In [34]:
# imorting Libaries
import pandas as pd
import numpy as np
from sklearn.linear_model import LogisticRegression
from  sklearn.linear_model import LogisticRegression
from  sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier







In [3]:
# Loading the data set to Pandas DataFrame
path = '/content/drive/MyDrive/DataSet/creditcard.csv'
credit_data = pd.read_csv(path)


In [4]:
# Shape of data i.e. total rows and columns
credit_data.shape


(284807, 31)

In [5]:
# Check all the columns
list[credit_data.columns]

list[Index(['Time', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10',
       'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20',
       'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount',
       'Class'],
      dtype='object')]

In [6]:
# display first 5 row
credit_data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [7]:
# display last 5 row  ( 172792 seconds = around 48 hours = 2 days transactions)
credit_data.tail()

Time         V1         V2        V3        V4        V5  \
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[5 rows x 31 columns]

In [8]:
# info ---- total memory consumned by dataframe  and type of data
credit_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [9]:
# Check the number of missing values in each column
credit_data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [10]:
# Checking the duplicates
credit_data.duplicated().sum()

1081

In [11]:
# Distribution of legit transactions & fraudulent transactions
credit_data['Class'].value_counts()

Class
0    284315
1       492
Name: count, dtype: int64

 From Above count, it shows there are 284,315 valid transactions and 492 fraudulent transactions in the dataset.


**This dataset is highly Unbalanced beacuse there is vaild transaction is much higher than frad transactions because of that we might face the issue in ML model**  

0----> Valid Transactions

1----> fraudulent transactions

In [12]:
# Separting the data for analysis
# here we are storing the valid transaction data in legit variable and fraudulent transactions in fraud

legit = credit_data[credit_data.Class == 0]
fraud = credit_data[credit_data.Class == 1]

In [13]:
print(legit.shape)
print(fraud.shape)

(284315, 31)
(492, 31)


In [14]:
# Statistical measure of the data
legit.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [15]:
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [16]:
## Very important
# Compare the values for both transactions
credit_data.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

# Under Sampling


Build a sample dataset containing similar distribution of valid transaction and fraudulent transactions


No of fraudulent transactions  = 492

In [17]:
legit_sample = legit.sample(n=492) # here we take the random 492 data (sample) from vaild transaction


Concatenating two DataFrames

axis =0 : values add by rows

axis =1 : values add by columns

In [18]:
new_credit_data = pd.concat([legit_sample, fraud], axis = 0)

In [19]:
new_credit_data.head()

Time        V1        V2        V3        V4        V5        V6  \
232848  147333.0 -1.473635 -2.655714 -2.041579 -1.194292  1.736163 -1.714825   
274880  166259.0 -0.664630  1.112438 -0.425395 -0.700663  0.632587 -0.761380   
254237  156637.0  1.721229 -0.242598 -1.928853  0.631073  0.198211 -1.337395   
136820   81884.0  0.419261  1.075894 -1.830565  0.304414  2.447749  3.657619   
28936    35243.0  1.532821 -1.176313 -0.315560 -1.878669 -0.532892  0.748214   

              V7        V8        V9  ...       V21       V22       V23  \
232848  1.367842 -0.322548 -1.754054  ...  0.532248  0.816642  1.220693   
274880  0.723007  0.314608  0.091973  ... -0.163845 -0.508937 -0.189949   
254237  0.623381 -0.528394  0.327548  ...  0.314101  0.755904 -0.217301   
136820 -1.350603 -1.821333 -0.738213  ... -1.374251 -0.025866 -0.070288   
28936  -1.158099  0.181918 -1.890330  ... -0.249585 -0.376210 -0.163380   

             V24       V25       V26       V27       V28  Amount  Class  
232848  0.686423 -0.890879  0.559319 -0.000712  0.343082  389.98      0  
274880 -0.853111 -0.042523 -0.069492 -0.000785  0.100507   30.00      0  
254237 -0.119197  0.263315 -0.119092 -0.025653  0.002306  183.42      0  
136820  0.847776  0.967597 -0.217355  0.061037  0.177827    1.00      0  
28936  -1.738797  0.488461 -0.086306  0.033771 -0.009493   25.00      0  

[5 rows x 31 columns]

In [20]:
new_credit_data.tail()

Time        V1        V2        V3        V4        V5        V6  \
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[5 rows x 31 columns]

In [21]:
new_credit_data["Class"].value_counts()

Class
0    492
1    492
Name: count, dtype: int64

Now we have the uniform data which is good for buliding the good ML Model.


In [22]:
new_credit_data.groupby("Class").mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      96722.597561  0.106745 -0.055922 -0.022884  0.008838  0.064532   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0     -0.022253  0.030034 -0.005014 -0.034762  ...  0.020234 -0.077787   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.055011  0.037165 -0.027775 -0.003461  0.011511  0.005724  0.015300   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       89.336829  
1      122.211321  

[2 rows x 30 columns]

Based on the above observatio ,we found that the mean values remain nearly unchanged(for Balanced and unbalaned data, mean is almost same)that means dataset nature is not changed.This step is important to check whether we have good sample or bad sample.
In case of bad sample , we should have received change in means.

Splitting the data into Features & Targets



In [25]:
# features
X = new_credit_data.drop(columns = 'Class', axis =1)
print(X.head())

            Time        V1        V2        V3        V4        V5        V6  \
232848  147333.0 -1.473635 -2.655714 -2.041579 -1.194292  1.736163 -1.714825   
274880  166259.0 -0.664630  1.112438 -0.425395 -0.700663  0.632587 -0.761380   
254237  156637.0  1.721229 -0.242598 -1.928853  0.631073  0.198211 -1.337395   
136820   81884.0  0.419261  1.075894 -1.830565  0.304414  2.447749  3.657619   
28936    35243.0  1.532821 -1.176313 -0.315560 -1.878669 -0.532892  0.748214   

              V7        V8        V9  ...       V20       V21       V22  \
232848  1.367842 -0.322548 -1.754054  ...  0.541382  0.532248  0.816642   
274880  0.723007  0.314608  0.091973  ... -0.169575 -0.163845 -0.508937   
254237  0.623381 -0.528394  0.327548  ...  0.271610  0.314101  0.755904   
136820 -1.350603 -1.821333 -0.738213  ...  0.758016 -1.374251 -0.025866   
28936  -1.158099  0.181918 -1.890330  ... -0.312899 -0.249585 -0.376210   

             V23       V24       V25       V26       V27       V28  

In [26]:
# Traget
Y = new_credit_data['Class']

In [27]:
print(Y)

232848    0
274880    0
254237    0
136820    0
28936     0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64


Spliting the data into Training and Test


In [29]:
X_train,X_test, Y_train, Y_test = train_test_split(X,Y, test_size=0.2, stratify=Y, random_state=2)

In [30]:
print(X.shape, X_train.shape, X_test.shape)

(984, 30) (787, 30) (197, 30)


**Model Training**

Logistics Regression Model as Y (target/dependent variable) is binary.

In [31]:
model = LogisticRegression()

In [32]:
# training the Logistic Regression Model with Training Data
model.fit(X_train, Y_train)


LogisticRegression()

**Model Evaluation**

Accuracy Score


In [35]:
# Accuracy on Training data

X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)


In [36]:
print('Accuracy on Tarining Data', training_data_accuracy)

Accuracy on Tarining Data 0.9440914866581956


If Accuracy Score is 0.75 or higher than this i.e. prediction is good.
The accuracy is calculated by comparing the predicted labels (y_pred) with the true labels (y_test).

In [37]:
# Accuracy on test data

X_test_prediction = model.predict(X_test)
testing_data_accuracy = accuracy_score(X_test_prediction, Y_test)





In [38]:
print('Accuracy test Data', testing_data_accuracy)

Accuracy test Data 0.934010152284264


Accuracy is good for test and training data that means Model is good.

If Accuracy test and traning has differrnt value i.e. model is underfitting or overfitting.